projects/solar/test/test_sense_mcp3427.c

```diff
@@ -0,0 +1,339 @@
+#include "data_store.h"
+#include "event_queue.h"
+#include "i2c.h"
+#include "interrupt.h"
+#include "log.h"
+#include "mcp3427_adc.h"
+#include "ms_test_helpers.h"
+#include "sense.h"
+#include "sense_mcp3427.h"
+#include "soft_timer.h"
+#include "solar_config.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_SENSED_CH1_VALUE 0x1337
+#define TEST_SENSED_CH2_VALUE 0xDEAD
+#define TEST_STORED_VALUE TEST_SENSED_CH1_VALUE  // only CH1 is used
+
+// there must be MAX_SOLAR_MCP3427 valid data points after and including this data point
+#define TEST_DATA_POINT DATA_POINT_VOLTAGE_1
+
+typedef enum TestMcp3427Event {
+  TEST_MCP3427_EVENT_TRIGGER = 0,
+  TEST_MCP3427_EVENT_READY,
+  NUM_TEST_MCP3427_EVENTS,
+} TestMcp3427Event;
+
+// if using multiple MCP3427s, overwrite the addr pins to avoid overlap
+static Mcp3427Settings s_test_mcp3427_settings = {
+  .sample_rate = MCP3427_SAMPLE_RATE_12_BIT,
+  .amplifier_gain = MCP3427_AMP_GAIN_1,
+  .conversion_mode = MCP3427_CONVERSION_MODE_CONTINUOUS,
+  .port = I2C_PORT_1,
+  .addr_pin_0 = MCP3427_PIN_STATE_LOW,
+  .addr_pin_1 = MCP3427_PIN_STATE_LOW,
+  .adc_data_trigger_event = TEST_MCP3427_EVENT_TRIGGER,
+  .adc_data_ready_event = TEST_MCP3427_EVENT_READY,
+};
+
+// all possible address combinations of MCP3427s
+static struct {
+  I2CPort port;
+  Mcp3427PinState addr_pin_0;
+  Mcp3427PinState addr_pin_1;
+} s_mcp3427_address_combos[] = {
+  // float & float is the same address as low & low, so we don't include it
+  { I2C_PORT_1, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_1, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_FLOAT },
+  { I2C_PORT_1, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_HIGH },
+  { I2C_PORT_1, MCP3427_PIN_STATE_FLOAT, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_1, MCP3427_PIN_STATE_FLOAT, MCP3427_PIN_STATE_HIGH },
+  { I2C_PORT_1, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_1, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_FLOAT },
+  { I2C_PORT_1, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_HIGH },
+  { I2C_PORT_2, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_2, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_FLOAT },
+  { I2C_PORT_2, MCP3427_PIN_STATE_LOW, MCP3427_PIN_STATE_HIGH },
+  { I2C_PORT_2, MCP3427_PIN_STATE_FLOAT, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_2, MCP3427_PIN_STATE_FLOAT, MCP3427_PIN_STATE_HIGH },
+  { I2C_PORT_2, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_LOW },
+  { I2C_PORT_2, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_FLOAT },
+  { I2C_PORT_2, MCP3427_PIN_STATE_HIGH, MCP3427_PIN_STATE_HIGH },
+};
+
+// Helper function to get SenseMcp3427Settings that cover the max number of MCP3427s
```

Comment:
I like the verbosity of this, but I think a simple nested for (and remove the wanted cases) would've been alot simpler ...